multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,35 +95,59 @@ func NewMemberClusterSetReconciler(client client.Client,
 func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	clusterSet := &mcv1alpha2.ClusterSet{}
 	err := r.Get(ctx, req.NamespacedName, clusterSet)
-	r.commonAreaLock.Lock()
-	defer r.commonAreaLock.Unlock()
+	var clusterSetNotFound bool
+	var clusterSetCreated bool
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return ctrl.Result{}, err
 		}
-		klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
-		if r.remoteCommonArea != nil {
-			if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
-				// MemberClusterAnnounce could be kept in the leader cluster, if antrea-mc-controller crashes after the failure.
-				// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
-				return ctrl.Result{}, fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+		clusterSetNotFound = true
+	}
+
+	processClusterSet := func() error {
+		r.commonAreaLock.Lock()
+		defer r.commonAreaLock.Unlock()
+
+		if clusterSetNotFound {
+			klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
+			if err := r.cleanUpResources(ctx, false); err != nil {
+				return err
 			}
-			r.remoteCommonArea.Stop()
-			r.remoteCommonArea = nil
-			r.clusterSetConfig = nil
-			r.clusterID = common.InvalidClusterID
-			r.clusterSetID = common.InvalidClusterSetID
+			return nil
 		}
-		return ctrl.Result{}, nil
-	}
 
-	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+
+		// Handle create or update
+
+		newLeader := clusterSet.Spec.Leaders[0]
+		if r.installedLeader.clusterID == newLeader.ClusterID && r.installedLeader.serverUrl == newLeader.Server &&
+			r.installedLeader.secretName == newLeader.Secret {
+			klog.V(2).InfoS("No change for leader cluster configuration")
+			return nil
+		}
+
+		var secretChangeOnly bool
+		if r.installedLeader.clusterID == newLeader.ClusterID && r.installedLeader.serverUrl == newLeader.Server && r.installedLeader.secretName != newLeader.Secret {
+			secretChangeOnly = true
+		}
+
+		if r.clusterID == common.InvalidClusterID {
+			clusterSetCreated = true
```

Comment:
This is very hard for me to understand. Even `r.clusterID != common.InvalidClusterID`, it is possible a new ClusterSet has been created, but just that cleanup of the old one is not done yet.

Could we just check `r.clusterID != newClusterID` and call `cleanUpResources` if true?